add parameter form docs (#17972)
PrefectHQ/prefect: docs/v3/tutorials/form-building.mdx

This allows you to:
- check runtime parameter values against the schema (from the UI or the SDK)
- build a user-friendly form in the Prefect UI
- easily reuse parameter types in similar workflows

In this tutorial, we'll craft a workflow signature that the Prefect UI will render as a self-documenting form.

## Motivation

Let's say you have a workflow that triggers a marketing email blast which looks like:

{/* pmd-metadata: notest */}
```python
@flow
def send_marketing_email(
    mailing_lists: list[str],
    subject: str,
    body: str,
    test_mode: bool = False,
    attachments: list[str] | None = None
):
    """
    Send a marketing email blast to the given lists.

    Args:
        mailing_lists: A list of lists to email.
        subject: The subject of the email.
        body: The body of the email.
        test_mode: Whether to send a test email.
        attachments: A list of attachments to include in the email.
    """
    ...
```

When you deploy this flow, Prefect will automatically inspect your function signature and generate a form for you:

![initial form](/v3/img/tutorials/parameters/0.png)

> This is good enough for many cases, but consider the additional practical constraints:

Reviewer: ```suggestion
This is good enough for many cases, but consider these additional constraints that could arise from business needs or tech stack restrictions:
```

- there are only a few valid values for `mailing_lists`
- the `subject` must not exceed 30 characters
- no more than 5 `attachments` are allowed

You _can_ simply check these constraints in the body of your flow function:

{/* pmd-metadata: notest */}
```python
@flow
def send_marketing_email(...):
    if len(subject) > 30:
        raise ValueError("Subject must be less than 30 characters")
    if mailing_lists not in ["newsletter", "customers", "beta-testers"]:
        raise ValueError("Invalid list to email")
    if len(attachments) > 5:
        raise ValueError("Too many attachments")
    
    # etc...
```

but there are several downsides to this:

> - you have spin up the infrastructure associated with your flow in order to check the constraints, which is wasteful if it turns out that bad parameters were provided

Reviewer: ```suggestion
- you have to spin up the infrastructure associated with your flow in order to check the constraints, which is wasteful if it turns out that bad parameters were provided
```

<Accordion title="Full example code so far">

{/* pmd-metadata: notest */}
```python
from typing import Literal

from prefect import flow
from pydantic import BaseModel, Field

MailingList = Literal["newsletter", "customers", "beta-testers"]

class EmailContent(BaseModel):
    subject: str = Field(max_length=30)
    body: str = Field(default=...)
    attachments: list[str] = Field(default_factory=list, max_length=5)

@flow
def send_marketing_email(
    mailing_list: list[MailingList],
    content: EmailContent,
    test_mode: bool = False,
):
    pass

if __name__ == "__main__":
    send_marketing_email.serve()
```

</Accordion>

### Using `json_schema_extra` to order fields in the form

By default, your flow parameters are rendered in the order defined by your `@flow` function signature.

Within a given `BaseModel` subclass, parameters are rendered in the following order:
- parameters with a `default` value are rendered first, alphabetically
- parameters without a `default` value are rendered next, alphabetically

You can control the order of the parameters within a `BaseModel` subclass by passing `json_schema_extra` to the `Field` constructor with a `position` key.

> Taking our `EmailContent` model from the previous example, let's enforce that `subject` should be `first`, then `body`, then `attachments`.

Reviewer: ```suggestion
Taking our `EmailContent` model from the previous example, let's enforce that `subject` should be displayed first, then `body`, then `attachments`.
```
